src/network/mod.rs

```diff
@@ -0,0 +1,175 @@
+use anyhow::{Context, Result};
+use async_trait::async_trait;
+use dusk_plonk::prelude::PublicParameters;
+use kate_recovery::{
+	config,
+	data::Cell,
+	matrix::{Dimensions, Position},
+};
+use mockall::automock;
+use sp_core::H256;
+use std::{sync::Arc, time::Duration};
+use tokio::time::Instant;
+use tracing::info;
+
+use crate::proof;
+
+pub mod p2p;
+pub mod rpc;
+
+#[async_trait]
+#[automock]
+pub trait Client {
+	async fn fetch_verified(
+		&self,
+		block_number: u32,
+		block_hash: H256,
+		dimensions: Dimensions,
+		commitments: &[[u8; config::COMMITMENT_SIZE]],
+		positions: &[Position],
+	) -> Result<(Vec<Cell>, Vec<Position>, FetchStats)>;
+}
+
+pub struct FetchStats {
+	pub dht_fetched: f64,
+	pub dht_fetched_percentage: f64,
+	pub rpc_fetched: f64,
+	pub dht_put_success_rate: Option<f64>,
+	pub dht_put_duration: Option<f64>,
+}
+
+impl FetchStats {
+	pub fn new(
+		total: usize,
+		dht_fetched: usize,
+		rpc_fetched: usize,
+		dht_stats: Option<(f64, Duration)>,
+	) -> Self {
+		FetchStats {
+			dht_fetched: dht_fetched as f64,
+			dht_fetched_percentage: dht_fetched as f64 / total as f64,
+			rpc_fetched: rpc_fetched as f64,
+			dht_put_success_rate: dht_stats.map(|(rate, _)| rate),
+			dht_put_duration: dht_stats.map(|(_, duration)| duration.as_secs_f64()),
+		}
+	}
+}
+
+struct DHTWithRPCFallbackClient {
+	p2p_client: p2p::Client,
+	rpc_client: rpc::Client,
+	pp: Arc<PublicParameters>,
+	disable_rpc: bool,
+}
+
+type Commitments = [[u8; config::COMMITMENT_SIZE]];
+
+#[async_trait]
+impl Client for DHTWithRPCFallbackClient {
+	async fn fetch_verified(
+		&self,
+		block_number: u32,
+		block_hash: H256,
+		dimensions: Dimensions,
+		commitments: &Commitments,
+		positions: &[Position],
+	) -> Result<(Vec<Cell>, Vec<Position>, FetchStats)> {
+		let (mut dht_fetched, mut unfetched) = self
+			.p2p_client
+			.fetch_cells_from_dht(block_number, positions)
+			.await;
+
+		let (verified, mut unverified) = proof::verify(
+			block_number,
+			dimensions,
+			&dht_fetched,
+			commitments,
+			self.pp.clone(),
+		)
+		.context("Failed to verify fetched cells")?;
+
+		info!(
```

Comment:
We should add a separate metric of the time it took to fetch & verify cells from DHT.